ydb/src/load_balancer/nearest_dc_balancer.rs

```diff
@@ -0,0 +1,388 @@
+                RandomLoadBalancer::new(),
+            )),
+        }
+    }
+}
+
+pub(crate) struct NearestDCBalancer {
+    discovery_state: Arc<DiscoveryState>,
+    state_sender: Sender<Arc<DiscoveryState>>,
+    ping_token: CancellationToken,
+    waiter: Arc<WaiterImpl>,
+    config: BalancerConfig,
+    balancer_state: Arc<Mutex<BalancerState>>,
+}
+
+impl NearestDCBalancer {
+    pub(crate) fn new(config: BalancerConfig) -> YdbResult<Self> {
+        let discovery_state = Arc::new(DiscoveryState::default());
+        let balancer_state = Arc::new(Mutex::new(BalancerState::default()));
+        let balancer_state_updater = balancer_state.clone();
+        let (state_sender, state_reciever) = watch::channel(discovery_state.clone());
+
+        let ping_token = CancellationToken::new();
+        let ping_token_clone = ping_token.clone();
+
+        let waiter = Arc::new(WaiterImpl::new());
+        let waiter_clone = waiter.clone();
+
+        tokio::spawn(async move {
+            Self::adjust_local_dc(
+                balancer_state_updater,
+                state_reciever,
+                ping_token_clone,
+                waiter_clone,
+            )
+            .await
+        });
+
+        Ok(Self {
+            discovery_state,
+            state_sender,
+            ping_token,
+            waiter,
+            config,
+            balancer_state,
+        })
+    }
+}
+
+impl Drop for NearestDCBalancer {
+    fn drop(&mut self) {
+        self.ping_token.cancel();
+    }
+}
+
+#[async_trait::async_trait]
+impl Waiter for NearestDCBalancer {
+    async fn wait(&self) -> YdbResult<()> {
+        self.waiter().wait().await
+    }
+}
+
+impl LoadBalancer for NearestDCBalancer {
+    fn endpoint(&self, service: Service) -> YdbResult<Uri> {
+        self.get_endpoint(service)
+    }
+
+    fn set_discovery_state(&mut self, discovery_state: &Arc<DiscoveryState>) -> YdbResult<()> {
+        match self.config.fallback_strategy.borrow_mut() {
+            FallbackStrategy::BalanceWithOther(balancer) => {
+                balancer.set_discovery_state(discovery_state)?
+            }
+            FallbackStrategy::Error => (),
+        }
+        self.discovery_state = discovery_state.clone();
+        let _ = self.state_sender.send(discovery_state.clone());
+        Ok(())
+    }
+
+    fn waiter(&self) -> Box<dyn Waiter> {
+        let self_waiter = Box::new(self.waiter.clone());
+        match self.config.fallback_strategy.borrow() {
+            FallbackStrategy::BalanceWithOther(balancer) => {
+                Box::new(AllWaiter::new(vec![self_waiter, balancer.waiter()]))
+            }
+            FallbackStrategy::Error => self_waiter,
+        }
+    }
+}
+
+pub(super) const NODES_PER_DC: usize = 5;
+pub(super) const PING_TIMEOUT_SECS: u64 = 60;
+
+impl NearestDCBalancer {
+    fn get_endpoint(&self, service: Service) -> YdbResult<Uri> {
+        match self.balancer_state.try_lock() {
```

Comment:
use lock instead of try_lock: we don`t need fast fail at the path: if code doesn't see endpoint it will return to retrier. The retraier will see custom error and fail transaction.

Better wait until update state finish: it is fast operation.

the method should return a error is the ballancer is empty

Reply:
Changed to `blocking_lock()`